extendr-api/src/robj/mod.rs

```diff
@@ -734,65 +734,125 @@ where
-    {
-        None
+        Self: 'a;
+}
+
+trait SExpAccessor {
+    const XP: u32;
+    type AccessorType;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType;
+}
+
+impl SExpAccessor for Rbool {
+    const XP: u32 = LGLSXP;
+    type AccessorType = raw::c_int;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { INTEGER(x) }
+    }
+}
+impl SExpAccessor for i32 {
+    const XP: u32 = INTSXP;
+    type AccessorType = raw::c_int;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { INTEGER(x) }
+    }
+}
+impl SExpAccessor for u32 {
+    const XP: u32 = INTSXP;
+    type AccessorType = raw::c_int;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { INTEGER(x) }
+    }
+}
+impl SExpAccessor for Rint {
+    const XP: u32 = INTSXP;
+    type AccessorType = raw::c_int;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { INTEGER(x) }
+    }
+}
+impl SExpAccessor for f64 {
+    const XP: u32 = REALSXP;
+    type AccessorType = Self;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { REAL(x) }
+    }
+}
+impl SExpAccessor for Rfloat {
+    const XP: u32 = REALSXP;
+    type AccessorType = f64;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { REAL(x) }
+    }
+}
+impl SExpAccessor for u8 {
+    const XP: u32 = RAWSXP;
+    type AccessorType = Rbyte;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { RAW(x) }
+    }
+}
+impl SExpAccessor for Rstr {
+    const XP: u32 = STRSXP;
+    type AccessorType = SEXP;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { STRING_PTR(x) }
+    }
+}
+impl SExpAccessor for c64 {
+    const XP: u32 = CPLXSXP;
+    type AccessorType = Rcomplex;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { COMPLEX(x) }
+    }
+}
+impl SExpAccessor for Rcplx {
+    const XP: u32 = CPLXSXP;
+    type AccessorType = Rcomplex;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { COMPLEX(x) }
+    }
+}
+impl SExpAccessor for Rcomplex {
+    const XP: u32 = CPLXSXP;
+    type AccessorType = Rcomplex;
+    fn acessor(x: SEXP) -> *mut Self::AccessorType {
+        unsafe { COMPLEX(x) }
     }
 }
 
-macro_rules! make_typed_slice {
-    ($type: ty, $fn: tt, $($sexp: tt),* ) => {
-        impl<'a> AsTypedSlice<'a, $type> for Robj
-        where
-            Self : 'a,
-        {
-            fn as_typed_slice(&self) -> Option<&'a [$type]> {
-                match self.sexptype() {
-                    $( $sexp )|* => {
-                        unsafe {
-                            let ptr = $fn(self.get()) as *const $type;
-                            Some(std::slice::from_raw_parts(ptr, self.len()))
-                        }
-                    }
-                    _ => None
-                }
+impl<'a, T> AsTypedSlice<'a, T> for Robj
+where
+    Self: 'a,
+    T: SExpAccessor,
+{
+    fn as_typed_slice(&self) -> Option<&'a [T]> {
```

Comment:
Since we're rewriting this, I again think we should consider if this should return a `Result`. One good reason is so we can add an error message.

Reply:
First, I need to get all these small changes and PRs through. Then I'm doing that. I did an MVP and it made sense, and you could do it, but to de-macrofi things as well, a little more care has to be dealt.
But it is on the agenda. I just have to survive this hellish loop of compilation issues.